Authorization Result changes

kreedula opened the pull request:
- Added Authorization Result implementation for AAD and MSTSS
- Changes to base Authorization result
- Need to address few Todo's after discussing with team
- Need to add unit tests to the change

heidijinxujia commented:
Travis failed because of the PMD violations. Could you have a look at it?
```Execution failed for task ':common:pmd'.
> 2 PMD rule violations were found.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/AuthorizationErrorResponse.java:
```diff
@@ -1,28 +1,72 @@
+    private String mError;
+    private String mErrorDescription;
+
+    public AuthorizationErrorResponse(String error, String errorDescription) {
```
Nit: Let's declare params `final` -- not going to call out every instance, but let's apply this everywhere

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationResult.java:
```diff
@@ -22,7 +22,117 @@
+            default:
+                setAuthorizationErrorResponse(AuthorizationStatus.FAIL,
+                        new AuthorizationErrorResponse(AuthorizationErrorResponse.UNKNOWN_ERROR,
+                                AuthorizationErrorResponse.UNKNOWN_RESULT_CODE + "[" + resultCode + "}"));
```
Nit: this log message contains a mix of `[` and `}` characters

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationResult.java:
```diff
@@ -22,7 +22,117 @@
+
+    private static final String TAG = AzureActiveDirectoryAuthorizationResult.class.getSimpleName();
+
+    public AzureActiveDirectoryAuthorizationResult(int resultCode, final Intent data) {
```
Let's javadoc this constructor + params

iambmelt replied:
Please restrict the functionality of constructors to setting fields (and optionally logging) _only_.

I'll save the rant and link you to [page 3 of Wolter, Ruffer, and Hevery's 'Writing Testable Code'](https://drive.google.com/file/d/1tXHR-0vzeyNcgzlzFjT1VroRkQqa56xt/view?usp=sharing). Chances are, this logic belongs elsewhere.

Consider using a factory or builder to create these objects, preferably using constructors which accept arguments specific to the way(s) this class needs to get initialized. Avoid calling `new` inside constructors, as this creates dependencies which can't be practically mocked. For instance, this constructor creates instances of `AuthorizationErrorResponse` -- this inter-class dependency isn't clear from this constructor's API and doesn't allow for testing to substitute alternative implementations of `AuthorizationErrorResponse`. Ideally, the relationship between these classes should be expressed as an interface like `IAuthorizationErrorResponse` in the event that `AuthorizationErrorResponse` needs to become `MicrosoftAuthorizationErrorResponse` or `AmazonAuthorizationErrorResponse` in the future

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationResult.java:
```diff
@@ -22,7 +22,117 @@
+                break;
+
+            case AuthenticationConstants.UIResponse.BROWSER_CODE_AUTHENTICATION_EXCEPTION:
+                //Serializable authException = extras.getSerializable(AuthenticationConstants.Browser.RESPONSE_AUTHENTICATION_EXCEPTION);
```
Remove commented-out code

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationResult.java:
```diff
@@ -22,7 +22,117 @@
+
+            case AuthenticationConstants.UIResponse.BROWSER_CODE_AUTHENTICATION_EXCEPTION:
+                //Serializable authException = extras.getSerializable(AuthenticationConstants.Browser.RESPONSE_AUTHENTICATION_EXCEPTION);
+                //TODO : AuthenticationException is not part of Common, what needs to be done here.
```
`Common` has its own `ClientException` and `ServiceException` classes -- take a look at those and determine which (if either) suits your needs and is semantically the most appropriate. If neither is, let's follow up

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationResult.java:
```diff
@@ -22,7 +22,117 @@
+import static com.microsoft.identity.common.adal.internal.AuthenticationConstants.Browser.RESPONSE_ERROR_MESSAGE;
+import static com.microsoft.identity.common.adal.internal.AuthenticationConstants.Browser.RESPONSE_FINAL_URL;
 
 public class AzureActiveDirectoryAuthorizationResult extends AuthorizationResult {
```
Consider creating and `extending` an intermediate subclass of `AuthorizationResult` called `MicrosoftAuthorizationResult`-- chances are, there's some opportunity for code reuse [or there will be in the future] between `AzureActiveDirectoryAuthorizationResult` and `MicrosoftStsAuthorizationResult`; that re-use should live in a Microsoft-specific superclass that exists separately of `AuthorizationResult`.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsAuthorizationResult.java:
```diff
@@ -22,7 +22,99 @@
+    public static final String ERROR_DESCRIPTION = "error_description";
+
+
+    public MicrosoftStsAuthorizationResult(int resultCode, final Intent data) {
```
See feedback for other class; same principles apply

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationResult.java:
```diff
@@ -22,7 +22,117 @@
+        switch (resultCode) {
+
+            case AuthenticationConstants.UIResponse.BROWSER_CODE_CANCEL:
+                Logger.verbose(TAG, null, "User cancel the request in webview: " + requestId);
```
Is this passing null as the `correlationId`? There's a two-arg constructor which uses the `correlationId` associated to the thread (if there is one....)

addev-ashish commented on common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/AuthorizationResult.java:
```diff
@@ -33,38 +33,53 @@
 
-        return null;
+    /**
+     * @return The {@link AuthorizationStatus} indicating the auth status for the request sent to authorize endopoint.
```
typo

addev-ashish commented on common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/AuthorizationResult.java:
```diff
@@ -33,38 +33,53 @@
+    }
+
+    /**
+     * @return The {@link AuthorizationResponse} indicating the auth response  for the request sent to authorize endopoint.
```
nit: copy-paste of endopoint, please fix here and everywhere else as well.

kreedula commented:
Thank you @iambmelt and @addev-ashish for your feedback. I have added factory pattern to create the AuthorizationResult objects. Also I have added specific subclasses of AuthorizationResponse, AuthorizationErrorResponse of the providers in this change. Please review it whenever you get a chance

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationResponse.java:
```diff
@@ -0,0 +1,47 @@
+import com.microsoft.identity.common.internal.providers.oauth2.AuthorizationResponse;
+
+
+public class AzureActiveDirectoryAuthorizationResponse extends AuthorizationResponse {
```
Shouldn't this now `extend MicrosoftAuthorizationResponse`?

If yes, consider pulling `setCorrelationId()` & `getCorrelationId()` into the superclass

heidijinxujia replied:
+1

kreedula replied:
Yes you're right. It's was unintentional. Thank you for pointing this out. Fixed it now

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationResponse.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public void setCorrelationId(final String correlationId) {
+        this.mCorrelationId = correlationId;
```
Can drop the use of `this`

kreedula replied:
Done.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationResult.java:
```diff
@@ -22,7 +22,20 @@
+public class AzureActiveDirectoryAuthorizationResult extends MicrosoftAuthorizationResult {
+
+
+    public AzureActiveDirectoryAuthorizationResult(final AuthorizationStatus authStatus, final AzureActiveDirectoryAuthorizationResponse authResponse) {
```
This constructor implementation is the same as `MicrosoftStsAuthorizationResult` -- consider pulling this up into `MicrosoftAuthorizationResult`

kreedula replied:
Good catch and suggestion, moved it super class.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationResult.java:
```diff
@@ -22,7 +22,20 @@
+        setAuthorizationResponse(authResponse);
+    }
+
+    public AzureActiveDirectoryAuthorizationResult(final AuthorizationStatus authStatus, final AuthorizationErrorResponse errorResponse) {
```
This constructor implementation is the same as `MicrosoftStsAuthorizationResult` -- consider pulling this up into `MicrosoftAuthorizationResult`

kreedula replied:
Done.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationResultFactory.java:
```diff
@@ -0,0 +1,132 @@
+            case AuthenticationConstants.UIResponse.BROWSER_CODE_AUTHENTICATION_EXCEPTION:
+                //TODO : Verify that a ClientException is serialized here after Broker Implementation
+                Serializable clientException = extras.getSerializable(AuthenticationConstants.Browser.RESPONSE_AUTHENTICATION_EXCEPTION);
+                if (clientException != null && clientException instanceof ClientException) {
```
`clientException` seems an assumptive name if we have to test
`(clientException instanceof ClientException)`

Let's call it something like `responseAuthenticationException` before we cast it. After the cast, this name seems fine

kreedula replied:
Done.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationResultFactory.java:
```diff
@@ -0,0 +1,132 @@
+    private static final String ERROR_CODES = "error_codes";
+
+    @Override
+    public AuthorizationResult createAuthorizationResult(final int resultCode, final Intent data) {
```
This is looking better -- can we lower the number of `return` statements?

Consider the following [contrived] example of adding logging to properties of our results -- multiple return statements increases our work, adding 4 new lines to the method for each `case`, instead of two.

Having a single return statement also makes it easier to set a breakpoint at the result and see what's coming out of a method

```java
public Fruit getProduce(Color color) {
    final Fruit result;

    switch(color) {
        case RED:
            result = new Apple();
            Log.info("Result weight: " + result.getWeight());
            Log.info("Result unit price: " + result.getPrice(Currency.US_DOLLARS));
            return result;

        case GREEN:
            result = new Pear();
            Log.info("Result weight: " + result.getWeight());
            Log.info("Result unit price: " + result.getPrice(Currency.US_DOLLARS));
            return result;
            
        default:
            throw new UnsupportedFruitException("Not an apple or a pear! Sad face!");
    }
}
```

versus
```java
public Fruit getProduce(Color color) {
    final Fruit result;

    switch(color) {
        case RED:
            result = new Apple();
            break;

        case GREEN:
            result = new Pear();
            break;

        default:
            throw new UnsupportedFruitException("Not an apple or a pear! Sad face!");
    }
    
    Log.info("Result weight: " + result.getWeight());
    Log.info("Result unit price: " + result.getPrice(Currency.US_DOLLARS));

    return result;
}
```

kreedula replied:
I like idea definitely see this super useful for debugging. Thanks for the suggestion. Made required changes.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationResultFactory.java:
```diff
@@ -0,0 +1,132 @@
+
+    private AzureActiveDirectoryAuthorizationResult parseUrlAndCreateAuthorizationResult(final String url) {
+        HashMap<String, String> urlParameters = StringExtensions.getUrlParameters(url);
+        if (urlParameters == null || urlParameters.isEmpty()) {
```
Style nit: I like to see a newline before conditionals to lower the density of what I'm reading (except when the conditional is the first line of a method). Kind of like paragraphs in a book.

This method hit my previous comment about multiple `return` statements -- not an code _problem_, but a readability/debugging choice

kreedula replied:
Done.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/AuthorizationResultFactory.java:
```diff
@@ -0,0 +1,28 @@
+ * Abstract Factory class to construct {@link AuthorizationResult}
+ */
+
+public abstract class AuthorizationResultFactory {
```
😄

kreedula replied:
Haha irony !. Edited to make it a bit more logical.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationResultFactory.java:
```diff
@@ -0,0 +1,132 @@
+    private AzureActiveDirectoryAuthorizationResult parseUrlAndCreateAuthorizationResult(final String url) {
+        HashMap<String, String> urlParameters = StringExtensions.getUrlParameters(url);
+        if (urlParameters == null || urlParameters.isEmpty()) {
+            Logger.warn(TAG, null, "Invalid server response, empty query string from the webview redirect.");
```
Consider using the two-arg `Logger.warn` call if we don't have access to the `correlationId` in this location

kreedula replied:
Done.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationResultFactory.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    private AzureActiveDirectoryAuthorizationResult parseUrlAndCreateAuthorizationResult(final String url) {
+        HashMap<String, String> urlParameters = StringExtensions.getUrlParameters(url);
```
Looks like this could be declared `final`

kreedula replied:
Done.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsAuthorizationResultFactory.java:
```diff
@@ -0,0 +1,95 @@
+    public static final String MSSTS_AUTHORIZATION_FINAL_URL = "com.microsoft.identity.client.finalUrl";
+
+    @Override
+    public AuthorizationResult createAuthorizationResult(final int resultCode, final Intent data) {
```
Another place where we could reduce the number of `return` statements for easier logging, debuggability, whatever

kreedula replied:
Done.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsAuthorizationResultFactory.java:
```diff
@@ -0,0 +1,95 @@
+    private MicrosoftStsAuthorizationResult createAuthorizationResultWithErrorResponse(final AuthorizationStatus authStatus,
+                                                                                       final String error,
+                                                                                       final String errorDescription) {
+        Logger.info(TAG, null, "Error is returned from webview redirect");
```
Consider using two-arg log statements if correlationId is null

kreedula replied:
Done.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsAuthorizationResultFactory.java:
```diff
@@ -0,0 +1,95 @@
+                    ErrorStrings.STATE_MISMATCH,
+                    MicrosoftAuthorizationErrorResponse.STATE_NOT_RETURNED);
+        } else {
+            //TODO : validate state
```
Not requesting a change, but some things to think about for the future:
- How might this method signature change once state validation is included?
- Will is throw an Exception? Checked or unchecked?
    * Who will catch this Exception and what can that class do to propagate the Exception back to the caller?
- Which class will be responsible for performing state validation? Will this logic be reusable?

kreedula replied:
Correct me if I am wrong but per my understanding we wouldn't be throwing an exception here. We would construct an `AuthorizationErrorResponse` with `MicrosoftAuthorizationErrorResponse.STATE_NOT_THE_SAME` and return back the `AuthorizationResult`. The calling class  should handle this error like any other result having `AuthorizationErrorMessage` and do the needful. 

As for which class would be responsible for state validation, I'm thinking of a resuable util method/class which compares the state from request and response. But this should not effect the structure of this class.

iambmelt replied:
Right on -- sounds good to me 👍

heidijinxujia commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsAuthorizationResponse.java:
```diff
@@ -0,0 +1,38 @@
+
+public class MicrosoftStsAuthorizationResponse extends AuthorizationResponse {
+
+    public MicrosoftStsAuthorizationResponse(final String code) {
```
State and state verification is required in both ADAL and MSAL. Please remove this API. If there is no state returned, we should throw error.

heidijinxujia commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsAuthorizationResponse.java:
```diff
@@ -0,0 +1,38 @@
+
+public class MicrosoftStsAuthorizationResponse extends AuthorizationResponse {
+
+    public MicrosoftStsAuthorizationResponse(final String code) {
```
State and state verification is required in both ADAL and MSAL. Please remove this API. If there is no state returned, we should throw error.

kreedula replied:
Thanks for pointing out. Removed from Microsoft specific classes

heidijinxujia commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsAuthorizationErrorResponse.java:
```diff
@@ -0,0 +1,11 @@
+package com.microsoft.identity.common.internal.providers.microsoft.microsoftsts;
```
MIssing copyright statement.

kreedula replied:
Done.

heidijinxujia commented on common/src/androidTest/java/com/microsoft/identity/common/AzureActiveDirectoryAuthorizationResultFactoryTest.java:
```diff
@@ -0,0 +1,236 @@
+package com.microsoft.identity.common;
```
Missing copyright statement.

kreedula replied:
Done.

heidijinxujia commented on common/src/androidTest/java/com/microsoft/identity/common/MicrosoftStsAuthorizationResultFactoryTest.java:
```diff
@@ -0,0 +1,166 @@
+package com.microsoft.identity.common;
```
Copyright missing.

kreedula replied:
Done.

heidijinxujia commented on common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/AuthorizationErrorResponse.java:
```diff
@@ -1,28 +1,59 @@
+    public void setErrorDescription(final String mErrorDescription) {
+        this.mErrorDescription = mErrorDescription;
+    }
+
```
nit: extra line

kreedula replied:
Done.

heidijinxujia commented on common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/AuthorizationErrorResponse.java:
```diff
@@ -1,28 +1,59 @@
+        mErrorDescription = errorDescription;
+    }
+
+    public String getError() {
```
Please add javadoc for all public API.

heidijinxujia commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsAuthorizationResultFactory.java:
```diff
@@ -0,0 +1,95 @@
+package com.microsoft.identity.common.internal.providers.microsoft.microsoftsts;
```
Missing copyright.

heidijinxujia commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsAuthorizationResultFactory.java:
```diff
@@ -0,0 +1,95 @@
+        return createAuthorizationResultWithErrorResponse(AuthorizationStatus.FAIL,
+                MicrosoftAuthorizationErrorResponse.AUTHORIZATION_FAILED,
+                MicrosoftAuthorizationErrorResponse.AUTHORIZATION_SERVER_INVALID_RESPONSE);
+
```
nit: extra line

heidijinxujia commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsAuthorizationResultFactory.java:
```diff
@@ -0,0 +1,95 @@
+            return createAuthorizationResultWithErrorResponse(AuthorizationStatus.FAIL,
+                    urlParameters.get(ERROR), urlParameters.get(ERROR_DESCRIPTION));
+        }
+        return createAuthorizationResultWithErrorResponse(AuthorizationStatus.FAIL,
```
nit: please add one line after }

heidijinxujia commented on common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/AuthorizationResultFactory.java:
```diff
@@ -0,0 +1,28 @@
+     * @return {@link AuthorizationResult}
+     */
+    public abstract AuthorizationResult createAuthorizationResult(final int resultCode, final Intent data);
+
```
nit: extra lines

heidijinxujia commented on common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/AuthorizationResultFactory.java:
```diff
@@ -0,0 +1,28 @@
+package com.microsoft.identity.common.internal.providers.oauth2;
```
copyright missing

kreedula replied:
Done.

heidijinxujia replied:
?

heidijinxujia commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/MicrosoftAuthorizationErrorResponse.java:
```diff
@@ -0,0 +1,31 @@
+package com.microsoft.identity.common.internal.providers.microsoft;
```
copyright missing.

kreedula replied:
Done.

heidijinxujia commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationResponse.java:
```diff
@@ -0,0 +1,47 @@
+
+    private String mCorrelationId;
+
+    public AzureActiveDirectoryAuthorizationResponse(final String code) {
```
ADAL and MSAL authorization response is `state` required. We should remove this constructor.

kreedula replied:
Done.

heidijinxujia commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationResponse.java:
```diff
@@ -0,0 +1,47 @@
+        super(code);
+    }
+
+    public AzureActiveDirectoryAuthorizationResponse(final String code, final String state) {
```
Please add javadoc for poublic apis.

kreedula replied:
Done.

kreedula commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationResultFactory.java:
```diff
@@ -0,0 +1,132 @@
+        }
+    }
+
+    private AzureActiveDirectoryAuthorizationResult validateAndCreateAuthorizationResult(final String code,
```
Done.

kreedula commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationResultFactory.java:
```diff
@@ -0,0 +1,132 @@
+            return createAuthorizationResultWithErrorResponse(AuthorizationStatus.FAIL,
+                    MicrosoftAuthorizationErrorResponse.AUTHORIZATION_FAILED,
+                    MicrosoftAuthorizationErrorResponse.AUTHORIZATION_SERVER_INVALID_RESPONSE);
+        }
```
Done.

kreedula commented:
Thanks @iambmelt and @heidijinxujia for your review and suggestions. I have addressed them via https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/133/commits/23311e60390387f8d2daa527b52283e5517fb6cf

@iambmelt Checkstyle and pmd are taken care of as well. Ran it locally to confirm

heidijinxujia commented on common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/AuthorizationStatus.java:
```diff
@@ -0,0 +1,27 @@
+package com.microsoft.identity.common.internal.providers.oauth2;
```
copyright?

kreedula replied:
Sorry I missed a couple again, added them now. Thanks @heidijinxujia

heidijinxujia commented on common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/OAuth2Strategy.java:
```diff
@@ -81,7 +81,7 @@ public GenericAuthorizationResponse requestAuthorization(
         AuthorizationResult result = authorizationStrategy.requestAuthorization(request); //NOPMD Suppressing PMD warning for unused variable
         //TODO: Reconcile authorization result and response
-        AuthorizationResponse response = new AuthorizationResponse();
+        AuthorizationResponse response = new AuthorizationResponse("");
```
We should have the AuthorizationResult consuming logic here. 
If AuthorizationResult has successful auth code returned,continue.
If failed, here AuthorizationErrorResponse returned, we throw the exception.

heidijinxujia replied:
You can make it a TODO and make a separate PR for this change.

kreedula replied:
Sure. I wasn't aware about this. I will push another PR for this. There is a TODO already here.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsAuthorizationResultFactory.java:
```diff
@@ -0,0 +1,133 @@
+                    MicrosoftAuthorizationErrorResponse.AUTHORIZATION_FAILED,
+                    MicrosoftAuthorizationErrorResponse.AUTHORIZATION_SERVER_INVALID_RESPONSE);
+        }
+        else if (urlParameters.containsKey(CODE)) {
```
Nit: This is super super super minor, but you wouldn't mind bringing the `else if` so it's formatted like:

```java
if (someCondition()) {
    // ...
} else if (someOtherCondition()) {
    // ...
}
```

kreedula replied:
My bad, I misunderstood your comment of having a new line after "}" conditionals for readability and applied the changes here as well going against the standard IDE formatting. I was hesitant to do that but thought may be it's a pattern we follow in the team here. Glad to change it back like before.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/AuthorizationErrorResponse.java:
```diff
@@ -1,28 +1,58 @@
+        return mError;
+    }
+
+    public void setError(final String mError) {
```
The input variable should just be `error`, since the `m` prefix denotes a member field.

We can lose the `this` syntax as well, since this name will no longer shadow the class' copy of `mError`

kreedula replied:
Generated getter/setter and missed to edit it. Thanks for spotting this.

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/AuthorizationErrorResponse.java:
```diff
@@ -1,28 +1,58 @@
+        return mErrorDescription;
+    }
+
+    public void setErrorDescription(final String mErrorDescription) {
```
Same feedback as above -- input variable should just be `errorDescription`. Can lose the `this`

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsAuthorizationResultFactory.java:
```diff
@@ -0,0 +1,133 @@
+            result = createAuthorizationResultWithErrorResponse(AuthorizationStatus.FAIL,
+                    urlParameters.get(ERROR), urlParameters.get(ERROR_DESCRIPTION));
+        }
+        else {
```
More nitpick stuff -- let's bring this up so it's like `} else {`

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsAuthorizationResultFactory.java:
```diff
@@ -0,0 +1,133 @@
+        else if (urlParameters.containsKey(CODE)) {
+            result = validateAndCreateAuthorizationResult(urlParameters.get(CODE), urlParameters.get(STATE));
+        }
+        else if (urlParameters.containsKey(ERROR)) {
```
Same here

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsAuthorizationResultFactory.java:
```diff
@@ -0,0 +1,133 @@
+            result = createAuthorizationResultWithErrorResponse(AuthorizationStatus.FAIL,
+                    MicrosoftAuthorizationErrorResponse.AUTHORIZATION_FAILED,
+                    MicrosoftAuthorizationErrorResponse.AUTHORIZATION_SERVER_INVALID_RESPONSE);
+        }
```
Newline after this closing `}`

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/AuthorizationResponse.java:
```diff
@@ -50,6 +50,24 @@
+     * @param code  The authorization code generated by the authorization server.
+     */
+    public AuthorizationResponse(final String code){
+        new AuthorizationResponse(code, null);
```
Can you spot the bug in this code? 😉

iambmelt replied:
Helpful hint: try declaring the fields `mCode` and `mState` as final and seeing what happens

kreedula replied:
Such an obvious bug, creating a separate instance which won't be used at all and the variables in intended class will not be initialized.  Sorry to have missing this 😞 . Thank you for spotting this @iambmelt. Fixed it

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsAuthorizationResultFactory.java:
```diff
@@ -0,0 +1,133 @@
+        AuthorizationResult result;
+        switch (resultCode) {
+            case AuthenticationConstants.UIResponse.BROWSER_CODE_CANCEL:
+                Logger.verbose(TAG, null, "User cancel the request in webview.");
```
Another place where the two-arg log method can be used

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/AuthorizationResultFactory.java:
```diff
@@ -0,0 +1,28 @@
+ * Abstract Factory class to construct {@link AuthorizationResult}
+ */
+
+public abstract class AuthorizationResultFactory {
```
We can squeeze better type safety out of this class using a generic template.

Consider the following declaration for this class

```java
public abstract class AuthorizationResultFactory<GenericAuthorizationResult extends AuthorizationResult> {
    // variables, javadocs and stuff...
    public abstract GenericAuthorizationResult createAuthorizationResult(final int resultCode, final Intent data);
}
```

This will allow you to remove the `instanceof` checks inside of your unit tests and hence, no more type casting

iambmelt commented on common/src/androidTest/java/com/microsoft/identity/common/AzureActiveDirectoryAuthorizationResultFactoryTest.java:
```diff
@@ -0,0 +1,236 @@
+        intent.putExtras(new Bundle());
+        AzureActiveDirectoryAuthorizationResultFactory factory = new AzureActiveDirectoryAuthorizationResultFactory();
+        AuthorizationResult result = factory.createAuthorizationResult(AuthenticationConstants.UIResponse.BROWSER_CODE_CANCEL, intent);
+        assertTrue(result instanceof AzureActiveDirectoryAuthorizationResult);
```
See https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/133/files/34372525a1342e7fc0e5ff69a49a0e871423275a#r195851360 -- we can avoid these checks/casts

kreedula replied:
Love it excellent suggestion @iambmelt  :ok_hand:. Made changes as required

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsAuthorizationResultFactory.java:
```diff
@@ -0,0 +1,95 @@
+
+import java.util.HashMap;
+
+public class MicrosoftStsAuthorizationResultFactory extends AuthorizationResultFactory {
```
See [this comment](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/133/files/34372525a1342e7fc0e5ff69a49a0e871423275a#r195851360) and consider the following class declaration

```java
public class MicrosoftStsAuthorizationResultFactory extends AuthorizationResultFactory<MicrosoftStsAuthorizationResult> {
    // ...
}
```

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectoryAuthorizationResultFactory.java:
```diff
@@ -0,0 +1,132 @@
+import java.io.Serializable;
+import java.util.HashMap;
+
+public class AzureActiveDirectoryAuthorizationResultFactory extends AuthorizationResultFactory {
```
See [this comment](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/133/files/34372525a1342e7fc0e5ff69a49a0e871423275a#r195851360) and consider the following declaration of this class

```java
public class AzureActiveDirectoryAuthorizationResultFactory extends AuthorizationResultFactory<AzureActiveDirectoryAuthorizationResult> {
    // ...
}
```

iambmelt commented on common/src/androidTest/java/com/microsoft/identity/common/AzureActiveDirectoryAuthorizationResultFactoryTest.java:
```diff
@@ -0,0 +1,258 @@
+    public void testBrowserCodeCancel() {
+        Intent intent = new Intent();
+        intent.putExtras(new Bundle());
+        AzureActiveDirectoryAuthorizationResultFactory factory = new AzureActiveDirectoryAuthorizationResultFactory();
```
Once this factory is genericized, you can change the left-side declaration to: `AuthorizationResultFactory<AzureActiveDirectoryAuthorizationResult>` for lower coupling inside of this test.

If you want to _really_ trick this out, you can declare the `AuthorizationResultFactory<AzureActiveDirectoryAuthorizationResult>` as a field in this class and then implement

```java
@Before
public void setUp() {
    mAuthorizationResultFactory = new AzureActiveDirectoryAuthorizationResultFactory();
}
```

That way, if we ever modify this constructor, we only have to update one line of code.

kreedula replied:
Makes sense 👍

iambmelt commented on common/src/androidTest/java/com/microsoft/identity/common/AzureActiveDirectoryAuthorizationResultFactoryTest.java:
```diff
@@ -0,0 +1,267 @@
+        assertNull(result.getAuthorizationErrorResponse());
+        assertEquals(AuthorizationStatus.SUCCESS, result.getAuthorizationStatus());
+        AuthorizationResponse response = result.getAuthorizationResponse();
+        assertTrue(response instanceof AzureActiveDirectoryAuthorizationResponse);
```
This is looking like another good opportunity for some added type-safety

iambmelt commented on common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/AuthorizationResult.java:
```diff
@@ -22,68 +22,50 @@
  * this class should also contain the AuthorizationResponse which contains the details returned from the
  * In the case of an error/exception this class should return the associated exception
  */
 public abstract class AuthorizationResult {
```
Here's another place we can use the generic type parameter to prevent users of this class from casting their results.

```java
public abstract class AuthorizationResult
        <GenericAuthorizationResponse extends AuthorizationResponse,
         GenericAuthorizationErrorResponse extends AuthorizationErrorResponse> {

    // Other fields omitted for brevity...

    private GenericAuthorizationResponse mAuthorizationResponse;
    private GenericAuthorizationErrorResponse mAuthorizationErrorResponse;

    // Other accesssors omitted for brevity...

    public GenericAuthorizationResponse getAuthorizationResponse() {
        return mAuthorizationResponse;
    }

    public GenericAuthorizationErrorResponse getAuthorizationErrorResponse() {
        return mAuthorizationErrorResponse;
    }

    protected void setAuthorizationErrorResponse(final GenericAuthorizationErrorResponse authErrorResponse) {
        mAuthorizationErrorResponse = authErrorResponse;
    }

    protected void setAuthorizationResponse(final GenericAuthorizationResponse authResponse) {
        mAuthorizationResponse = authResponse;
    }
}
```

Subclasses like `MicrosoftAuthorizationResult` can then apply this pattern to more specific subclasses:

>Note: I would consider making `MicrosoftAuthorizationResult` `abstract`

```java
public class MicrosoftAuthorizationResult
        <GenericMicrosoftAuthorizationResponse extends MicrosoftAuthorizationResponse,
                GenericMicrosoftAuthorizationErrorResponse extends MicrosoftAuthorizationErrorResponse>
        extends AuthorizationResult<MicrosoftAuthorizationResponse, MicrosoftAuthorizationErrorResponse> {

    public MicrosoftAuthorizationResult(final AuthorizationStatus authStatus, final GenericMicrosoftAuthorizationResponse authResponse) {
        setAuthorizationStatus(authStatus);
        setAuthorizationResponse(authResponse);
    }

    public MicrosoftAuthorizationResult(final AuthorizationStatus authStatus, final GenericMicrosoftAuthorizationErrorResponse errorResponse) {
        setAuthorizationStatus(authStatus);
        setAuthorizationErrorResponse(errorResponse);
    }

}
```

Same idea for `AzureActiveDirectoryAuthorizationResult`

```java
public class AzureActiveDirectoryAuthorizationResult
        extends MicrosoftAuthorizationResult<AzureActiveDirectoryAuthorizationResponse, AzureActiveDirectoryAuthorizationErrorResponse>
```

And for `MicrosoftStsAuthorizationResult`

```java
public class MicrosoftStsAuthorizationResult
        extends MicrosoftAuthorizationResult<MicrosoftStsAuthorizationResponse, MicrosoftStsAuthorizationErrorResponse>
```

By doing this, subclasses will automatically enforce that all parameters are the proper-type. Users of these classes will never have to cast or `instanceof` their results, since the return-type is made explicit through the template

iambmelt replied:
I've left out the details of the accessor methods to save space/typing

iambmelt commented on common/src/androidTest/java/com/microsoft/identity/common/MicrosoftStsAuthorizationResultFactoryTest.java:
```diff
@@ -0,0 +1,196 @@
+        assertNull(result.getAuthorizationErrorResponse());
+        assertEquals(AuthorizationStatus.SUCCESS, result.getAuthorizationStatus());
+        AuthorizationResponse response = result.getAuthorizationResponse();
+        assertTrue(response instanceof MicrosoftStsAuthorizationResponse);
```
Can be removed, once the superclass is genericized

kreedula commented:
@iambmelt Thanks for the suggestion. I added type safety to the classes in this commit https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/133/commits/e6759a696727227e37791bb315085f7bc978feb6

iambmelt commented:
Suggested a few edits #139 (PR to this PR) -- I'd suggest looking at that and, if it merges, correcting the `checkstyle` errors. After that, I think we're good.

```
$ ./gradlew checkstyle
:common:checkstyle[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\dto\Account.java:392:5: Missing a Javadoc comment. [JavadocMethod]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\MicrosoftAuthorizationErrorResponse.java:33:5: Missing a Javadoc comment. [JavadocVariable]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\MicrosoftAuthorizationErrorResponse.java:35:5: Missing a Javadoc comment. [JavadocVariable]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\MicrosoftAuthorizationErrorResponse.java:37:5: Missing a Javadoc comment. [JavadocVariable]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\MicrosoftAuthorizationErrorResponse.java:39:5: Missing a Javadoc comment. [JavadocVariable]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\MicrosoftAuthorizationErrorResponse.java:41:5: Missing a Javadoc comment. [JavadocVariable]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\MicrosoftAuthorizationErrorResponse.java:43:5: Missing a Javadoc comment. [JavadocVariable]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\MicrosoftAuthorizationErrorResponse.java:45:5: Missing a Javadoc comment. [JavadocVariable]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\MicrosoftAuthorizationErrorResponse.java:47:5: Missing a Javadoc comment. [JavadocVariable]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\MicrosoftAuthorizationErrorResponse.java:49:5: Missing a Javadoc comment. [JavadocVariable]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\MicrosoftAuthorizationErrorResponse.java:51:5: Missing a Javadoc comment. [JavadocVariable]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\azureactivedirectory\AzureActiveDirectoryAuthorizationErrorResponse.java:56:44: 'mErrorCodes' hides a field. [HiddenField]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\azureactivedirectory\AzureActiveDirectoryAuthorizationResultFactory.java:34:8: Unused import - com.microsoft.identity.common.internal.providers.oauth2.AuthorizationErrorResponse. [UnusedImports]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\azureactivedirectory\AzureActiveDirectoryAuthorizationResultFactory.java:35:8: Unused import - com.microsoft.identity.common.internal.providers.oauth2.AuthorizationResult. [UnusedImports]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\azureactivedirectory\AzureActiveDirectoryAuthorizationResultFactory.java:43: First sentence should end with a period. [JavadocStyle]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\azureactivedirectory\AzureActiveDirectoryAuthorizationResultFactory.java:65: switch without "default" clause. [MissingSwitchDefault]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\microsoftsts\MicrosoftStsAuthorizationResultFactory.java:32:8: Unused import - com.microsoft.identity.common.internal.providers.oauth2.AuthorizationErrorResponse. [UnusedImports]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\microsoftsts\MicrosoftStsAuthorizationResultFactory.java:39: First sentence should end with a period. [JavadocStyle]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\microsoftsts\MicrosoftStsAuthorizationResultFactory.java:48:5: Missing a Javadoc comment. [JavadocVariable]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\microsoftsts\MicrosoftStsAuthorizationResultFactory.java:89:25: ',' is not followed by whitespace. [WhitespaceAfter]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\microsoftsts\MicrosoftStsAuthorizationResultFactory.java:90:28: ',' is not followed by whitespace. [WhitespaceAfter]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\microsoftsts\MicrosoftStsAuthorizationResultFactory.java:100:29: ',' is not followed by whitespace. [WhitespaceAfter]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\microsoft\microsoftsts\MicrosoftStsOAuth2Strategy.java:60:53: ';' is not followed by whitespace. [WhitespaceAfter]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\oauth2\AuthorizationErrorResponse.java:43:5: Missing a Javadoc comment. [JavadocMethod]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\oauth2\AuthorizationErrorResponse.java:47:5: Missing a Javadoc comment. [JavadocMethod]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\oauth2\AuthorizationErrorResponse.java:51:5: Missing a Javadoc comment. [JavadocMethod]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\oauth2\AuthorizationErrorResponse.java:55:5: Missing a Javadoc comment. [JavadocMethod]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\oauth2\AuthorizationResponse.java:57:52: '{' is not preceded with whitespace. [WhitespaceAround]
[ant:checkstyle] [WARN] C:\Development\azure-activedirectory-library-for-android\common\common\src\main\java\com\microsoft\identity\common\internal\providers\oauth2\AuthorizationResponse.java:66:72: '{' is not preceded with whitespace. [WhitespaceAround]
```

iambmelt commented:
@kreedula The template change in #139 was my bad 😬

kreedula commented:
Thank you @iambmelt Addresed minor format and checkstyle issues in the latest commit https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/133/commits/c863d8ce4b47730e13ade52c86af02f33bf6a6bf

iambmelt commented:
Thank you, @kreedula!

heidijinxujia commented:
#127
